File: leetcode/java/pascalTriangle.java
/*
 * given numRows, generate the first numRows of Pascal's triangle
 * e.g. numRows=5, return:
 * 1
 * 1,1
 * 1,2,1
 * 1,3,3,1
 * 1,4,6,4,1
 * */

import java.io.*;
import java.util.*;

public class pascalTriangle{
    public ArrayList<ArrayList<Integer>> generate(int nRows){
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        if(nRows < 1)    return res;
        ArrayList<Integer> lst = new ArrayList<Integer>();
        lst.add(1);
        res.add(lst);
        for(int i=1;i<nRows;++i){
            ArrayList<Integer> last = res.get(i-1);
            ArrayList<Integer> now = new ArrayList<Integer>();
            now.add(last.get(0));
            for(int j=1;j<i;++j){
                now.add(last.get(j-1) + last.get(j));
            }
            now.add(last.get(i-1));
            res.add(now);
        }
        return res;
    }

    void show(ArrayList<ArrayList<Integer>> tri){
        for(ArrayList<Integer> arr : tri){
            for(Integer i : arr){
                System.out.print(i + " ");
            }
            System.out.println();
        }
    }
    void test_01(){
        Scanner sc = new Scanner(System.in);
        while(true){
            System.out.println("please input string: ");
            String s = sc.nextLine().trim();
            if(s.isEmpty())        break;
            int n = Integer.parseInt(s);
            show(generate(n));    
        }
    }
    public static void main(String[] args){
        pascalTriangle pt = new pascalTriangle();
        pt.test_01();
    }
}

File: leetcode/src/pascalTriangle.cpp
/*
 * given numRows, generate the first numRows of Pascal's triangle
 * e.g. given nRows=5, return below:
 * 1
 * 1,1
 * 1,2,1
 * 1,3,3,1
 * 1,4,6,4,1
 * */

#include <vector>

using namespace std;

class Solution{

public:
    vector<vector<int> > generate(int nRows){
        vector<vector<int> > res;
        if(nRows < 1)    return res;

        vector<int> base;
        base.push_back(1);
        res.push_back(base);

        for(int i = 1; i < nRows; ++i){ // i is row index, and row i has i+1 elements
            vector<int> vec;
            vec.push_back(res[i-1][0]);

            for(int j = 1; j < i; ++j){ // row i-1 has one element less than row i
                vec.push_back(res[i-1][j-1] + res[i-1][j]);
            }

            vec.push_back(res[i-1][i-1]);

            res.push_back(vec);
        }

        return res;
    }
};

/* unit test is in ../cpp_unittest/pascalTriangle_unittest */
